File: codes/php/Array.filter.php
<?php

$array_filter_v1 = function ($callback_function, $an_array) {
    // JavaScript-like Array.filter() function
    $filtered_data = [];
    foreach ($an_array as $array_item_index => $array_item) {
        $is_condition_match = $callback_function($array_item, $array_item_index, $an_array);
        if ($is_condition_match) array_push($filtered_data, $array_item);
    }
    return $filtered_data;
};

$array_filter_v2 = function ($callback_function, $an_array) {
    // JavaScript-like Array.filter() function
    $filtered_data = [];
    foreach ($an_array as $array_item_index => $array_item) {
        $is_condition_match = $callback_function($array_item, $array_item_index, $an_array);
        if ($is_condition_match) $filtered_data = [...$filtered_data, $array_item];
    }
    return $filtered_data;
};

$array_filter_v3 = function ($callback_function, $an_array) {
    // JavaScript-like Array.filter() function
    $filtered_data = [];
    foreach ($an_array as $array_item_index => $array_item) {
        if ($callback_function($array_item, $array_item_index, $an_array)) array_push($filtered_data, $array_item);
    }
    return $filtered_data;
};

$array_filter_v4 = function ($callback_function, $an_array) {
    // JavaScript-like Array.filter() function
    $filtered_data = [];
    foreach ($an_array as $array_item_index => $array_item) {
        if ($callback_function($array_item, $array_item_index, $an_array)) $filtered_data = [...$filtered_data, $array_item];
    }
    return $filtered_data;
};

echo("\n\n// JavaScript-like Array.filter() in PHP Array");

$numbers = [12, 34, 27, 23, 65, 93, 36, 87, 4, 254];
echo("\nnumbers: " . str_replace(",", ", ", json_encode($numbers)));

echo("\n// using JavaScript-like Array.filter() function \"array_filter_v1\"");

$even_numbers_only = $array_filter_v1(fn($number) => $number % 2 === 0, $numbers);
echo("\neven numbers only: " . str_replace(",", ", ", json_encode($even_numbers_only)));
// even numbers only: [12,34,36,4,254]

$odd_numbers_only = $array_filter_v1(fn($number) => $number % 2 !== 0, $numbers);
echo("\nodd numbers only: " . str_replace(",", ", ", json_encode($odd_numbers_only)));
// odd numbers only: [27,23,65,93,87]

echo("\n// using JavaScript-like Array.filter() function \"array_filter_v2\"");

$even_numbers_only = $array_filter_v2(fn($number) => $number % 2 === 0, $numbers);
echo("\neven numbers only: " . str_replace(",", ", ", json_encode($even_numbers_only)));
// even numbers only: [12,34,36,4,254]

$odd_numbers_only = $array_filter_v2(fn($number) => $number % 2 !== 0, $numbers);
echo("\nodd numbers only: " . str_replace(",", ", ", json_encode($odd_numbers_only)));
// odd numbers only: [27,23,65,93,87]

echo("\n// using JavaScript-like Array.filter() function \"array_filter_v3\"");

$even_numbers_only = $array_filter_v3(fn($number) => $number % 2 === 0, $numbers);
echo("\neven numbers only: " . str_replace(",", ", ", json_encode($even_numbers_only)));
// even numbers only: [12,34,36,4,254]

$odd_numbers_only = $array_filter_v3(fn($number) => $number % 2 !== 0, $numbers);
echo("\nodd numbers only: " . str_replace(",", ", ", json_encode($odd_numbers_only)));
// odd numbers only: [27,23,65,93,87]

echo("\n// using JavaScript-like Array.filter() function \"array_filter_v4\"");

$even_numbers_only = $array_filter_v4(fn($number) => $number % 2 === 0, $numbers);
echo("\neven numbers only: " . str_replace(",", ", ", json_encode($even_numbers_only)));
// even numbers only: [12,34,36,4,254]

$odd_numbers_only = $array_filter_v4(fn($number) => $number % 2 !== 0, $numbers);
echo("\nodd numbers only: " . str_replace(",", ", ", json_encode($odd_numbers_only)));
// odd numbers only: [27,23,65,93,87]

echo("\n// using PHP Array.filter() built-in function \"array_filter\"");

$even_numbers_only = array_filter($numbers, fn($number) => $number % 2 === 0);
echo("\neven numbers only: " . str_replace(",", ", ", json_encode($even_numbers_only)));
// even numbers only: [12,34,36,4,254]

$odd_numbers_only = array_filter($numbers, fn($number) => $number % 2 !== 0);
echo("\nodd numbers only: " . str_replace(",", ", ", json_encode($odd_numbers_only)));
// odd numbers only: [27,23,65,93,87]

echo("\n\n// JavaScript-like Array.filter() in PHP Associative Array");

$products = [
    [
        "code" => "pasta",
        "price" => 321
    ],
    [
        "code" => "bubble_gum",
        "price" => 233
    ],
    [
        "code" => "potato_chips",
        "price" => 5
    ],
    [
        "code" => "towel",
        "price" => 499
    ]
];
echo("\nproducts: " . json_encode($products, JSON_PRETTY_PRINT));

echo("\n// using JavaScript-like Array.filter() function \"array_filter_v1\"");

$products_with_price_less_than_or_equal_to_100_only = $array_filter_v1(fn($product) => $product['price'] <= 100, $products);
echo("\nproducts with price <= 100 only: " . json_encode($products_with_price_less_than_or_equal_to_100_only, JSON_PRETTY_PRINT));
// products with price <= 100 only: [
//     {
//         "code": "potato_chips",
//         "price": 5
//     }
// ]

$products_with_price_more_than_or_equal_to_100_only = $array_filter_v1(fn($product) => $product['price'] >= 100, $products);
echo("\nproducts with price >= 100 only: " . json_encode($products_with_price_more_than_or_equal_to_100_only, JSON_PRETTY_PRINT));
// products with price >= 100 only: [
//     {
//         "code": "pasta",
//         "price": 321
//     },
//     {
//         "code": "bubble_gum",
//         "price": 233
//     },
//     {
//         "code": "towel",
//         "price": 499
//     }
// ]

echo("\n// using JavaScript-like Array.filter() function \"array_filter_v2\"");

$products_with_price_less_than_or_equal_to_100_only = $array_filter_v2(fn($product) => $product['price'] <= 100, $products);
echo("\nproducts with price <= 100 only: " . json_encode($products_with_price_less_than_or_equal_to_100_only, JSON_PRETTY_PRINT));
// products with price <= 100 only: [
//     {
//         "code": "potato_chips",
//         "price": 5
//     }
// ]

$products_with_price_more_than_or_equal_to_100_only = $array_filter_v2(fn($product) => $product['price'] >= 100, $products);
echo("\nproducts with price >= 100 only: " . json_encode($products_with_price_more_than_or_equal_to_100_only, JSON_PRETTY_PRINT));
// products with price >= 100 only: [
//     {
//         "code": "pasta",
//         "price": 321
//     },
//     {
//         "code": "bubble_gum",
//         "price": 233
//     },
//     {
//         "code": "towel",
//         "price": 499
//     }
// ]

echo("\n// using JavaScript-like Array.filter() function \"array_filter_v3\"");

$products_with_price_less_than_or_equal_to_100_only = $array_filter_v3(fn($product) => $product['price'] <= 100, $products);
echo("\nproducts with price <= 100 only: " . json_encode($products_with_price_less_than_or_equal_to_100_only, JSON_PRETTY_PRINT));
// products with price <= 100 only: [
//     {
//         "code": "potato_chips",
//         "price": 5
//     }
// ]

$products_with_price_more_than_or_equal_to_100_only = $array_filter_v3(fn($product) => $product['price'] >= 100, $products);
echo("\nproducts with price >= 100 only: " . json_encode($products_with_price_more_than_or_equal_to_100_only, JSON_PRETTY_PRINT));
// products with price >= 100 only: [
//     {
//         "code": "pasta",
//         "price": 321
//     },
//     {
//         "code": "bubble_gum",
//         "price": 233
//     },
//     {
//         "code": "towel",
//         "price": 499
//     }
// ]

echo("\n// using JavaScript-like Array.filter() function \"array_filter_v4\"");

$products_with_price_less_than_or_equal_to_100_only = $array_filter_v4(fn($product) => $product['price'] <= 100, $products);
echo("\nproducts with price <= 100 only: " . json_encode($products_with_price_less_than_or_equal_to_100_only, JSON_PRETTY_PRINT));
// products with price <= 100 only: [
//     {
//         "code": "potato_chips",
//         "price": 5
//     }
// ]

$products_with_price_more_than_or_equal_to_100_only = $array_filter_v4(fn($product) => $product['price'] >= 100, $products);
echo("\nproducts with price >= 100 only: " . json_encode($products_with_price_more_than_or_equal_to_100_only, JSON_PRETTY_PRINT));
// products with price >= 100 only: [
//     {
//         "code": "pasta",
//         "price": 321
//     },
//     {
//         "code": "bubble_gum",
//         "price": 233
//     },
//     {
//         "code": "towel",
//         "price": 499
//     }
// ]

echo("\n// using PHP Array.filter() built-in function \"array_filter\"");

$products_with_price_less_than_or_equal_to_100_only = array_filter($products, fn($product) => $product['price'] <= 100);
echo("\nproducts with price <= 100 only: " . json_encode($products_with_price_less_than_or_equal_to_100_only, JSON_PRETTY_PRINT));
// products with price <= 100 only: [
//     {
//         "code": "potato_chips",
//         "price": 5
//     }
// ]

$products_with_price_more_than_or_equal_to_100_only = array_filter($products, fn($product) => $product['price'] >= 100);
echo("\nproducts with price >= 100 only: " . json_encode($products_with_price_more_than_or_equal_to_100_only, JSON_PRETTY_PRINT));
// products with price >= 100 only: [
//     {
//         "code": "pasta",
//         "price": 321
//     },
//     {
//         "code": "bubble_gum",
//         "price": 233
//     },
//     {
//         "code": "towel",
//         "price": 499
//     }
// ]

File: codes/perl/Array.filter.pl
use strict;
use warnings;
use JSON;
use Scalar::Util qw(looks_like_number);

# There's no JavaScript-like Array.filter() in Perl.
# But, we can create our own function to mimic it in Perl.

sub pretty_array_of_primitives {
    my $number_of_parameters = @_;
    my $result = "[";
    for (my $array_item_index = 0; $array_item_index < $number_of_parameters; $array_item_index += 1) {
        my $array_item = $_[$array_item_index];
        my $is_string = defined($array_item) && $array_item =~ /[0-9a-zA-Z`~!@#%&_=;':", \(\)\[\]\{\}\|\*\+\?\^\$\/\\\<\>\.\-]/;
        my $is_number = looks_like_number($array_item);
        last if (!$is_string && !$is_number);
        $result = $result . "\"" . $array_item . "\"" if ($is_string);
        $result = $result . $array_item if ($is_number);
        $result = $result . ", " if (($array_item_index + 1) != $number_of_parameters);
    }
    $result = $result . "]";
    return $result;
}

sub array_filter_v1 {
    # JavaScript-like Array.filter() function
    my ($callback_function, @an_array) = @_;
    my @data_filtered = ();
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        my $is_condition_match = $callback_function->($array_item, $array_item_index, \@an_array);
        if ($is_condition_match) {
            push(@data_filtered, $array_item);
        }
    }
    return @data_filtered;
}

sub array_filter_v2 {
    # JavaScript-like Array.filter() function
    my ($callback_function, @an_array) = @_;
    my @data_filtered = ();
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        my $is_condition_match = $callback_function->($array_item, $array_item_index, \@an_array);
        push(@data_filtered, $array_item) if ($is_condition_match);
    }
    return @data_filtered;
}

sub array_filter_v3 {
    # JavaScript-like Array.filter() function
    my ($callback_function, @an_array) = @_;
    my @data_filtered = ();
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        if ($callback_function->($array_item, $array_item_index, \@an_array)) {
            push(@data_filtered, $array_item);
        }
    }
    return @data_filtered;
}

sub array_filter_v4 {
    # JavaScript-like Array.filter() function
    my ($callback_function, @an_array) = @_;
    my @data_filtered = ();
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        push(@data_filtered, $array_item) if ($callback_function->($array_item, $array_item_index, \@an_array));
    }
    return @data_filtered;
}

print("\n# JavaScript-like Array.filter() in Perl Array\n");

my @numbers = (12, 34, 27, 23, 65, 93, 36, 87, 4, 254);
print("numbers: ", pretty_array_of_primitives(@numbers), "\n");

my @numbers_even;
my @numbers_odd;

print("# using JavaScript-like Array.filter() function \"array_filter_v1\"\n");

@numbers_even = array_filter_v1(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even numbers only: ", pretty_array_of_primitives(@numbers_even), "\n");
# even numbers only: [12, 34, 36, 4, 254]

@numbers_odd = array_filter_v1(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd numbers only: ", pretty_array_of_primitives(@numbers_odd), "\n");
# odd numbers only: [27, 23, 65, 93, 87]

print("# using JavaScript-like Array.filter() function \"array_filter_v2\"\n");

@numbers_even = array_filter_v2(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even numbers only: ", pretty_array_of_primitives(@numbers_even), "\n");
# even numbers only: [12, 34, 36, 4, 254]

@numbers_odd = array_filter_v2(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd numbers only: ", pretty_array_of_primitives(@numbers_odd), "\n");
# odd numbers only: [27, 23, 65, 93, 87]

print("# using JavaScript-like Array.filter() function \"array_filter_v3\"\n");

@numbers_even = array_filter_v3(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even numbers only: ", pretty_array_of_primitives(@numbers_even), "\n");
# even numbers only: [12, 34, 36, 4, 254]

@numbers_odd = array_filter_v3(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd numbers only: ", pretty_array_of_primitives(@numbers_odd), "\n");
# odd numbers only: [27, 23, 65, 93, 87]

print("# using JavaScript-like Array.filter() function \"array_filter_v4\"\n");

@numbers_even = array_filter_v4(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even numbers only: ", pretty_array_of_primitives(@numbers_even), "\n");
# even numbers only: [12, 34, 36, 4, 254]

@numbers_odd = array_filter_v4(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd numbers only: ", pretty_array_of_primitives(@numbers_odd), "\n");
# odd numbers only: [27, 23, 65, 93, 87]

print("\n# JavaScript-like Array.filter() in Perl Array of Hashes\n");

my @products = (
    {
        "code" => "pasta",
        "price" => 321
    },
    {
        "code" => "bubble_gum",
        "price" => 233
    },
    {
        "code" => "potato_chips",
        "price" => 5
    },
    {
        "code" => "towel",
        "price" => 499
    }
);

print("products: ", JSON->new->allow_nonref->pretty->encode(\@products));

my @products_below_100;
my @products_above_100;

print("# using JavaScript-like Array.filter() function \"array_filter_v1\"\n");

@products_below_100 = array_filter_v1(sub { my ($product) = @_; return $product->{"price"} <= 100; }, @products);
print("products with price <= 100 only: ", JSON->new->allow_nonref->pretty->encode(\@products_below_100));
# products with price <= 100 only: [
#     {
#         "code": "potato_chips",
#         "price": 5
#     }
# ]

@products_above_100 = array_filter_v1(sub { my ($product) = @_; return $product->{"price"} >= 100; }, @products);
print("products with price >= 100 only: ", JSON->new->allow_nonref->pretty->encode(\@products_above_100));
# products with price >= 100 only: [
#     {
#         "code": "pasta",
#         "price": 321
#     },
#     {
#         "code": "bubble_gum",
#         "price": 233
#     },
#     {
#         "code": "towel",
#         "price": 499
#     }
# ]

print("# using JavaScript-like Array.filter() function \"array_filter_v2\"\n");

@products_below_100 = array_filter_v2(sub { my ($product) = @_; return $product->{"price"} <= 100; }, @products);
print("products with price <= 100 only: ", JSON->new->allow_nonref->pretty->encode(\@products_below_100));
# products with price <= 100 only: [
#     {
#         "code": "potato_chips",
#         "price": 5
#     }
# ]

@products_above_100 = array_filter_v2(sub { my ($product) = @_; return $product->{"price"} >= 100; }, @products);
print("products with price >= 100 only: ", JSON->new->allow_nonref->pretty->encode(\@products_above_100));
# products with price >= 100 only: [
#     {
#         "code": "pasta",
#         "price": 321
#     },
#     {
#         "code": "bubble_gum",
#         "price": 233
#     },
#     {
#         "code": "towel",
#         "price": 499
#     }
# ]

print("# using JavaScript-like Array.filter() function \"array_filter_v3\"\n");

@products_below_100 = array_filter_v3(sub { my ($product) = @_; return $product->{"price"} <= 100; }, @products);
print("products with price <= 100 only: ", JSON->new->allow_nonref->pretty->encode(\@products_below_100));
# products with price <= 100 only: [
#     {
#         "code": "potato_chips",
#         "price": 5
#     }
# ]

@products_above_100 = array_filter_v3(sub { my ($product) = @_; return $product->{"price"} >= 100; }, @products);
print("products with price >= 100 only: ", JSON->new->allow_nonref->pretty->encode(\@products_above_100));
# products with price >= 100 only: [
#     {
#         "code": "pasta",
#         "price": 321
#     },
#     {
#         "code": "bubble_gum",
#         "price": 233
#     },
#     {
#         "code": "towel",
#         "price": 499
#     }
# ]

print("# using JavaScript-like Array.filter() function \"array_filter_v4\"\n");

@products_below_100 = array_filter_v4(sub { my ($product) = @_; return $product->{"price"} <= 100; }, @products);
print("products with price <= 100 only: ", JSON->new->allow_nonref->pretty->encode(\@products_below_100));
# products with price <= 100 only: [
#     {
#         "code": "potato_chips",
#         "price": 5
#     }
# ]

@products_above_100 = array_filter_v4(sub { my ($product) = @_; return $product->{"price"} >= 100; }, @products);
print("products with price >= 100 only: ", JSON->new->allow_nonref->pretty->encode(\@products_above_100));
# products with price >= 100 only: [
#     {
#         "code": "pasta",
#         "price": 321
#     },
#     {
#         "code": "bubble_gum",
#         "price": 233
#     },
#     {
#         "code": "towel",
#         "price": 499
#     }
# ]
